File: node/src/webhooks.ts
import {
  createHmac,
  timingSafeEqual,
} from 'node:crypto';
import {
  LinopayWebhookVerificationError,
} from './errors.js';

/**
 * Webhook signature scheme — HMAC-SHA256, version-prefixed.
 *
 * This is the mirror implementation of LinoPay's outbound-webhook
 * signer (C#: `shared/LinoPay.Application/Abstractions/WebhookSignature.cs`).
 * A merchant who receives a webhook verifies it by:
 *
 *   1. Reading the `X-LinoPay-Timestamp` and `X-LinoPay-Signature`
 *      headers.
 *   2. Constructing the signed payload `${timestamp}.${rawBody}`
 *      (timestamp is signed AS WELL AS sent — changing either
 *      invalidates the signature, which kills replay attacks).
 *   3. Recomputing HMAC-SHA256(secret, payload) and comparing the
 *      hex digest to the `v1=...` value in `X-LinoPay-Signature` in
 *      CONSTANT TIME (a byte-by-byte early return leaks the position
 *      of the first mismatch — enough to recover a signature).
 *
 * The constant-time compare is on purpose: Node's `crypto.timingSafeEqual`
 * runs in fixed time regardless of input length, which is what we want.
 *
 * Replay window: 5 minutes. A delivery older than that should be
 * rejected even if the signature itself checks out (mirrors
 * `WebhookSignature.ReplayWindow` upstream).
 */
export const SIGNATURE_HEADER = 'X-LinoPay-Signature';
export const TIMESTAMP_HEADER = 'X-LinoPay-Timestamp';
export const SCHEME_VERSION = 'v1';
export const REPLAY_WINDOW_SECONDS = 300;

export interface VerifyInput {
  /**
   * The raw HTTP request body as a string (NOT a re-serialised JSON
   * object — re-serialising changes whitespace and key ordering, both
   * of which break the signature).
   */
  body: string;
  /**
   * The `X-LinoPay-Timestamp` header value, as a Unix-seconds string.
   */
  timestampHeader: string;
  /**
   * The `X-LinoPay-Signature` header value (`v1=<hex>`).
   */
  signatureHeader: string;
  /**
   * The shared signing secret (`whsec_...`).
   */
  secret: string;
  /**
   * Override the default 5-minute replay window. Tests use this; so
   * does anything pulling a "now" from a controlled clock.
   */
  nowSeconds?: number;
}

/**
 * Verifies a webhook delivery. Throws
 * {@link LinopayWebhookVerificationError} on any mismatch; returns
 * `undefined` on success (a `boolean` would be more conventional but
 * the throw carries the *reason*, which is the data a merchant's own
 * handler logs — and we don't want them constructing their own message).
 */
export function verifyWebhook(input: VerifyInput): void {
  if (!input.signatureHeader || input.signatureHeader.trim() === '') {
    throw new LinopayWebhookVerificationError(
      'no-signature',
      'No signature header was supplied.',
    );
  }
  const ts = Number.parseInt(input.timestampHeader, 10);
  if (!Number.isFinite(ts)) {
    throw new LinopayWebhookVerificationError(
      'no-signature',
      'Timestamp header is missing or not a Unix-seconds integer.',
    );
  }
  if (input.secret === '') {
    throw new LinopayWebhookVerificationError(
      'wrong-secret',
      'A non-empty signing secret is required to verify a webhook.',
    );
  }

  // Signature check FIRST, then replay-window check. A merchant who
  // forges a signature deserves a `wrong-secret` reason regardless of
  // the timestamp freshness; the replay check is a defence against an
  // attacker replaying a *valid* signature, not against forging one.
  const expected = computeSignature(input.secret, ts, input.body);

  // Constant-time compare. Node's timingSafeEqual requires equal-length
  // buffers — pad the shorter side so a wrong-length candidate can't
  // crash the comparison and isn't short-circuited either.
  const a = Buffer.from(expected, 'utf8');
  let b = Buffer.from(input.signatureHeader, 'utf8');
  if (b.length !== a.length) {
    // Construct a same-length junk buffer for the safe-equal. We
    // already know this is a mismatch, but we need to spend the
    // constant time so we don't leak length info via timing.
    b = Buffer.alloc(a.length);
  }
  if (!timingSafeEqual(a, b)) {
    throw new LinopayWebhookVerificationError(
      'wrong-secret',
      'Webhook signature did not match.',
    );
  }

  const now = input.nowSeconds ?? Math.floor(Date.now() / 1000);
  if (Math.abs(now - ts) > REPLAY_WINDOW_SECONDS) {
    throw new LinopayWebhookVerificationError(
      'replay-out-of-window',
      `Timestamp ${ts} is outside the ${REPLAY_WINDOW_SECONDS}-second replay window (now=${now}).`,
    );
  }
}

/**
 * Computes the signing payload `${timestamp}.${body}` and returns
 * `v1=<lowercase hex HMAC-SHA256>`. Exported so a merchant's own test
 * suite can verify against the same byte-exact construction the SDK
 * uses for verification — recreating the spec from the verifier is a
 * classic source of integration bugs.
 */
export function computeSignature(secret: string, unixSeconds: number, body: string): string {
  if (secret === '') {
    throw new Error('computeSignature requires a non-empty secret.');
  }
  const payload = `${unixSeconds.toString(10)}.${body}`;
  const mac = createHmac('sha256', secret).update(payload, 'utf8').digest('hex');
  return `${SCHEME_VERSION}=${mac}`;
}

File: dotnet/src/LinoPay.Sdk/Webhooks.cs
using System.Security.Cryptography;
using System.Text;

namespace Linopay.Sdk;

/// <summary>
/// Webhook signature scheme — HMAC-SHA256, version-prefixed.
///
/// <para>
/// This is the mirror implementation of LinoPay's outbound-webhook
/// signer (C#:
/// <c>lime-payments/shared/LinoPay.Application/Abstractions/WebhookSignature.cs</c>).
/// A merchant who receives a webhook verifies it by:
/// </para>
/// <list type="number">
///   <item>Reading the <c>X-LinoPay-Timestamp</c> and
///   <c>X-LinoPay-Signature</c> headers.</item>
///   <item>Constructing the signed payload <c>${timestamp}.${rawBody}</c>
///   (timestamp is signed AS WELL AS sent — changing either
///   invalidates the signature, which kills replay attacks).</item>
///   <item>Recomputing HMAC-SHA256(secret, payload) and comparing the
///   hex digest to the <c>v1=...</c> value in <c>X-LinoPay-Signature</c>
///   in CONSTANT TIME (a byte-by-byte early return leaks the position
///   of the first mismatch — enough to recover a signature).</item>
/// </list>
/// </summary>
public static class WebhookSignature
{
    /// <summary>Header carrying <c>v1=&lt;hex&gt;</c>.</summary>
    public const string SignatureHeader = "X-LinoPay-Signature";

    /// <summary>Header carrying the Unix-seconds timestamp that was signed.</summary>
    public const string TimestampHeader = "X-LinoPay-Timestamp";

    /// <summary>Scheme version prefix on the signature value.</summary>
    public const string Version = "v1";

    /// <summary>Maximum age (in seconds) a delivery may be before a
    /// merchant should reject it. Advisory — enforced by the merchant,
    /// not by us.</summary>
    public static readonly TimeSpan ReplayWindow = TimeSpan.FromSeconds(300);

    /// <summary>
    /// Builds the byte-exact payload that gets HMAC'd:
    /// <c>${timestamp}.${body}</c>.
    /// </summary>
    public static string BuildSigningPayload(long unixSeconds, string body) =>
        unixSeconds.ToString(System.Globalization.CultureInfo.InvariantCulture) + "." + body;

    /// <summary>
    /// Computes the <c>X-LinoPay-Signature</c> value for a delivery:
    /// <c>"v1=" + lowercase hex HMAC-SHA256</c>.
    /// </summary>
    public static string Compute(string secret, long unixSeconds, string body)
    {
        if (string.IsNullOrEmpty(secret))
            throw new LinopayWebhookVerificationException("wrong-secret",
                "A non-empty signing secret is required to compute a webhook signature.");
        var payload = BuildSigningPayload(unixSeconds, body);
        var keyBytes = Encoding.UTF8.GetBytes(secret);
        var dataBytes = Encoding.UTF8.GetBytes(payload);
        var mac = HMACSHA256.HashData(keyBytes, dataBytes);
        return Version + "=" + Convert.ToHexString(mac).ToLowerInvariant();
    }

    /// <summary>
    /// Verifies a candidate signature in constant time. Throws
    /// <see cref="LinopayWebhookVerificationException"/> on any mismatch.
    /// Returns silently on success.
    /// </summary>
    public static void Verify(string secret, long unixSeconds, string body, string? candidate, long? nowSeconds = null)
    {
        if (string.IsNullOrWhiteSpace(candidate))
            throw new LinopayWebhookVerificationException("no-signature",
                "No signature header was supplied.");

        if (secret.Length == 0)
            throw new LinopayWebhookVerificationException("wrong-secret",
                "A non-empty signing secret is required to verify a webhook.");

        // Signature check FIRST, then replay-window check. A merchant
        // who forges a signature deserves a wrong-secret reason
        // regardless of the timestamp freshness; the replay check is a
        // defence against an attacker replaying a *valid* signature,
        // not against forging one.
        var expected = Compute(secret, unixSeconds, body);

        var a = Encoding.UTF8.GetBytes(expected);
        byte[] bRaw = Encoding.UTF8.GetBytes(candidate!);
        if (bRaw.Length != a.Length)
        {
            // Construct a same-length junk buffer for the safe-equal.
            // We already know this is a mismatch, but we need to spend
            // the constant time so we don't leak length info via timing.
            bRaw = new byte[a.Length];
        }

        if (!CryptographicOperations.FixedTimeEquals(a, bRaw))
            throw new LinopayWebhookVerificationException("wrong-secret",
                "Webhook signature did not match.");

        var now = nowSeconds ?? DateTimeOffset.UtcNow.ToUnixTimeSeconds();
        if (Math.Abs(now - unixSeconds) > (long)ReplayWindow.TotalSeconds)
            throw new LinopayWebhookVerificationException("replay-out-of-window",
                $"Timestamp {unixSeconds} is outside the {(int)ReplayWindow.TotalSeconds}-second replay window (now={now}).");
    }

    /// <summary>
    /// The user-facing wrapper: takes the convenience
    /// <see cref="WebhookVerifyArgs"/> record, throws on any
    /// mismatch, returns silently on success. Mirrors Phase 1's
    /// <c>webhooks.verify(input)</c>.
    /// </summary>
    public static void Verify(WebhookVerifyArgs args)
    {
        Verify(args.Secret, args.UnixSeconds, args.Body, args.Candidate, args.NowSeconds);
    }
}

public sealed record WebhookVerifyArgs(
    string Secret,
    long UnixSeconds,
    string Body,
    string? Candidate,
    long? NowSeconds = null);
